src/txn/txn.c

```diff
@@ -2308,7 +2308,22 @@ __wt_txn_rollback(WT_SESSION_IMPL *session, const char *cfg[])
 #endif
 
     if (cursor != NULL) {
-        WT_TRET(cursor->close(cursor));
+        /*
+         * Technically the WiredTiger API allows closing a cursor to return rollback. This is a
+         * strange error to get in the rollback path so we swallow that error here. Analysis made at
+         * the time suggests that it is impossible for resetting a cursor to return rollback, which
+         * is called from cursor close, but we cannot guarantee it.
+         *
+         * Because swallowing an error that you believe cannot happen doesn't make a lot of sense we
+         * assert the error is not generated in diagnostic mode.
+         */
+#ifdef HAVE_DIAGNOSTIC
+        int ret2 = cursor->close(cursor);
```

Comment:
Should we rather check for some pre-conditions that make `cursor->close()` never return `WT_ROLLBACK`, then assert that it doesn't do it?
Like, "there's no active transaction at this point"? Can we make this assertion outside of "if"?

Or if there's an error, then there could be an active transaction? If there's still one, then `close()` can potentially return a "rollback" (in future)?

What makes this call to `cursor->close()` unique so that it doesn't return `WT_ROLLBACK` compared to other ones? That's somehow implied here but is not checked.

I'm just trying to minimise the impact of this hidden dependency between implementations of `close()` and this check here. I'm worried that a potential mismatch might be a rare case, then it's a trap that's hard to find.

Alternatively, that kind of check could be done inside of `close()`. That way, we could also guarantee that given certain conditions, `close()` doesn't return `WT_ROLLBACK`.

Reply:
> Should we rather check for some pre-conditions that make cursor->close() never return WT_ROLLBACK, then assert that it doesn't do it?

We can't make that check as there aren't any. The only guarantee we have is the analysis I performed on WT-11871. I have tried to think of some but yeah, drawing blanks here.

> What makes this call to cursor->close() unique so that it doesn't return WT_ROLLBACK compared to other ones? 

Realistically I don't think any cursor->close() calls can return rollback given the investigation performed. It isn't special but this one is in the rollback path so it is "special" in that sense.

> I'm worried that a potential mismatch might be a rare case, then it's a trap that's hard to find.

If there exists a very rare case that returns rollback the #have_diagnostic check is intended to catch it.

> Alternatively, that kind of check could be done inside of close(). That way, we could also guarantee that given certain conditions, close() doesn't return WT_ROLLBACK.

This is the difficult part about this ticket I think, we don't want to guarantee that we can't return rollback which creates the scenario we have in `__wt_session_rollback`. If we were okay with guaranteeing it then I'd put the check in `__cursor_reset` instead and therefor neither `cursor->close()` nor `cursor->reset()` could return `WT_ROLLBACK`.

There's no good solution available here if we don't want to make guarantees that we _won't_ return certain error codes.

Reply:
Would it be correct to claim that WT can only return a `WT_ROLLBACK` only if there is an active transaction? Otherwise, if there's no transaction, this error doesn't make any sense to the user.

I understand that internally it's provided by a complex logic but we could also enforce it by an API level check.

By the way, this kind of guarantee should be apply to any API call - not only `rollback()` and/or `reset()`. What do you think?

Reply:
This was discussed on the ticket, I'm resolving it.